java/src/main/native/src/RmmJni.cpp

```diff
@@ -266,55 +274,217 @@ void set_java_device_memory_resource(
   }
 }
 
+/**
+ * @brief Resource that provides logging of allocations and frees in CSV
+ * format
+ */
+class logging_resource_adaptor final : public rmm::mr::device_memory_resource {
+ public:
+
+ logging_resource_adaptor(device_memory_resource* upstream, std::ostream * out, bool close_at_end): 
+     upstream(upstream), out(out), close_at_end(close_at_end) {}
+
+ /**
+  * @brief Return pointer to the upstream resource.
+  *
+  * @return pointer to the upstream resource.
+  */
+ device_memory_resource* get_wrapped_resource() {
+  return upstream;
+ }
+
+ const device_memory_resource* get_wrapped_resource() const {
+  return upstream;
+ }
+
+ /**
+  * @brief Checks whether the upstream resource supports streams.
+  *
+  * @return true The upstream resource supports streams
+  * @return false The upstream resource does not support streams.
+  */
+ bool supports_streams() const noexcept override {
+   return upstream->supports_streams();
+ }
+
+ /**
+  * @brief Query whether the resource supports the get_mem_info API.
+  * 
+  * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+  */
+ bool supports_get_mem_info() const noexcept override { return upstream->supports_streams(); }
+
+ virtual ~logging_resource_adaptor() {
+  if (close_at_end) {
+    delete out;
```

Comment:
Since we're (conditionally) deleting `out`, it would be safer to include/disallow a copy constructor and assignment operator, to protect against double deletes:
```cpp
private:
    logging_resource_adaptor(logging_resource_adaptor const&) = delete;
    logging_resource_adaptor& operator = (logging_resource_adaptor const&) = delete;
```

The generated defaults wouldn't be safe.

Reply:
Argh. We might also need to disable the move constructor, if we don't need it. Otherwise, the defaults will replicate `close_at_end`, and cause a double `delete`.